### PR TITLE
fix(client): nullability for *OrThrow methods via fluent API

### LINKED ISSUE #20183: Wrong nullable types with fluent API in Prisma 5.0
### Bug description

We use Fluent API very often and it works very well.
Since Prisma version 5.0, results for related models queried with fluent API are nullable, although `findUniqueOrThrow` was used on the base model and the relation field is not optional.

### How to reproduce

1. Create two models with a 1:n relation:
```prisma
model PlanGroup {
  id             Int                     @id @default(autoincrement()) @db.UnsignedInt
  plans          BuildingSitePlan[]
}

model Plan {
  id              Int         @id @default(autoincrement()) @db.UnsignedInt
  groupId         Int         @map("group_id") @db.UnsignedInt
  group           PlanGroup   @relation(fields: [groupId], references: [id], onDelete: Cascade)
}
```
3. Use Fluent API to query them:
```ts
(
    await prisma.plan
      .findUniqueOrThrow({
        where: {
          id,
        },
      })
      .group({
        select: {
          id: true,
        },
        
      })
  ).id
```
6. See typescript error ts(2531) "Object is possibly `null`"

### Expected behavior

In versions <5.0.0, queries for non-nullable related models with fluent api were non-nullable.

### Prisma information

```prisma
model PlanGroup {
  id             Int                     @id @default(autoincrement()) @db.UnsignedInt
  plans          BuildingSitePlan[]
}

model Plan {
  id              Int         @id @default(autoincrement()) @db.UnsignedInt
  groupId         Int         @map("group_id") @db.UnsignedInt
  group           PlanGroup   @relation(fields: [groupId], references: [id], onDelete: Cascade)
}
```

```ts
(
    await prisma.plan
      .findUniqueOrThrow({
        where: {
          id,
        },
      })
      .group({
        select: {
          id: true,
        },
        
      })
  ).id
```

### Environment & setup

- OS: macOS Ventura 13.2
- Database: MariaDB 10.11.3
- Node.js version: v20.2.0

### Prisma Version

```
prisma                  : 5.0.0
@prisma/client          : 5.0.0
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 6b0aef69b7cdfc787f822ecd7cdc76d5f1991584 (at node_modules/.pnpm/@prisma+engines@5.0.0/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 6b0aef69b7cdfc787f822ecd7cdc76d5f1991584 (at node_modules/.pnpm/@prisma+engines@5.0.0/node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 4.17.0-26.6b0aef69b7cdfc787f822ecd7cdc76d5f1991584
Default Engines Hash    : 6b0aef69b7cdfc787f822ecd7cdc76d5f1991584
Studio                  : 0.487.0
```

### PR DESCRIPTION
**Pull request by millsp**
closes https://github.com/prisma/prisma/issues/20183

### CODE REVIEW

File: packages/client/src/generation/TSClient/Model.ts [hunk 1]
```diff
@@ -555,13 +555,13 @@ ${f.name}<T extends ${getFieldArgName(f, name)}<ExtArgs> = {}>(args?: Subset<T,
         name,
       )}<ExtArgs>>): ${getReturnType({
         name: fieldTypeName,
-        actionName: f.outputType.isList ? DMMF.ModelAction.findMany : DMMF.ModelAction.findUnique,
+        actionName: f.outputType.isList ? DMMF.ModelAction.findMany : DMMF.ModelAction.findUniqueOrThrow,
```
Comment (millsp) [inline packages/client/src/generation/TSClient/Model.ts:558]: This is the main part of the fix. The previous code basically said, chaining will depend on `findUnique`, which in turn always adds ` | null` to any result. But we want null or only be present if the root operation is `findUnique` or `findFirst`, or if using `*OrThrow` and chaining on nullable/optional relations.

File: packages/client/src/generation/utils.ts [hunk 2]
```diff
@@ -248,28 +248,31 @@ export function getReturnType({
...
     }${promiseClose}`
   }
 
+  if (isChaining && actionName === 'findUniqueOrThrow') {
+    return `Prisma__${name}Client<${getType(
+      `$Types.GetResult<${name}Payload<ExtArgs>, T, '${actionName}'>`,
+      isList,
+    )} | ${isNullable ? 'null' : 'Null'}, ${isNullable ? 'null' : 'Null'}, ExtArgs>`
+  }
+
```
Comment (millsp) [inline packages/client/src/generation/utils.ts:261]: This is the code from below, moved up and handles `findUniqueOrThrow` operations when chaining. These happen when we generate types for chainable members. It also adds `null` for optional relations, and forwards any previous nullability on from the chain via `Null`. (see other [comment](https://github.com/prisma/prisma/pull/20197/files#r1261636730)).

File: packages/client/src/runtime/core/types/Extensions.ts [hunk 3]
```diff
@@ -149,7 +149,7 @@ type DynamicModelExtensionFluentApi<TypeMap extends TypeMapDef, M extends Proper
   [K in keyof TypeMap['model'][M]['payload']['objects']]:
     <A>(args?: Exact<A, Path<TypeMap['model'][M]['operations'][P]['args']['select'], [K]>>) =>
       & PrismaPromise<Path<DynamicModelExtensionFnResultBase<TypeMap, M, { select: { [P in K]: A } }, P>, [K]> | Null>
-      & DynamicModelExtensionFluentApi<TypeMap, (TypeMap['model'][M]['payload']['objects'][K] & {})['name'], P, Null>
+      & DynamicModelExtensionFluentApi<TypeMap, (TypeMap['model'][M]['payload']['objects'][K] & {})['name'], P, Null | Select<TypeMap['model'][M]['payload']['objects'][K], null>>
```
Comment (millsp) [inline packages/client/src/runtime/core/types/Extensions.ts:152]: Extensions suffered from a part of the problem, `null`s were not forwarded when chaining on optional relations.

File: packages/client/tests/functional/fluent-api-null/tests.ts [hunk 4]
```diff
@@ -225,6 +225,56 @@ testMatrix.setupTestSuite(() => {
...
+            id: '123',
+          },
+        })
+        .parent()
+        .children()
+
+      expectTypeOf<Awaited<typeof result>>().toEqualTypeOf<Child[] | null>()
+    })
+
+    test('findFirstOrThrow with optional to-one relation circling back to to-many relation', () => {
```
Comment (SevInf) [inline packages/client/tests/functional/fluent-api-null/tests.ts:266]: This test is identical to a previous one.

**Review by SevInf — APPROVED** [review]

**Pull request merged**